File: kvdom-dom/src/wasm32Main/kotlin/lt/petuska/kvdom/dom/node/Document.kt
package lt.petuska.kvdom.dom.node

import kotlinx.wasm.jsinterop.*

/**
 * https://developer.mozilla.org/en-US/docs/Web/API/Document
 */
actual class Document(arena: Arena, index: Object) : Node(arena, index) {
    /**
     * https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById
     */
    actual fun getElementById(id: String): Element? {
        return allocateArena().let { newArena ->
            js_Document_getElementById(
                arena, index,
                stringPointer(id),
                stringLengthBytes(id),
                newArena
            ).let {
                if (it < 0) {
                    null.also {
                        freeArena(newArena)
                    }
                } else {
                    Element(newArena, it)
                }
            }
        }
    }

    /**
     * https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement
     */
    actual fun createElement(tagName: String): Element {
        return allocateArena().let { newArena ->
            js_Document_createElement(
                arena, index,
                stringPointer(tagName),
                stringLengthBytes(tagName),
                newArena
            ).let {
                Element(newArena, it)
            }
        }
    }

    /**
     * https://developer.mozilla.org/en-US/docs/Web/API/Document/createTextNode
     */
    actual fun createTextNode(data: String): Text {
        return allocateArena().let { newArena ->
            js_Document_createTextNode(
                arena, index,
                stringPointer(data),
                stringLengthBytes(data),
                newArena
            ).let {
                Text(newArena, it)
            }
        }
    }

    /**
     * https://developer.mozilla.org/en-US/docs/Web/API/Document/defaultView
     */
    actual val defaultView: Window?
        get() = js_Document_defaultView(arena, index).let {
            if (it < 0) {
                null
            } else {
                Window(arena, js_Document_defaultView(arena, index))
            }
        }
}

@SymbolName("kvdom_Document_getElementById")
private external fun js_Document_getElementById(
    arena: Arena,
    index: Object,
    idPtr: Pointer,
    idLen: Int,
    resultArena: Arena
): Object

@SymbolName("kvdom_Document_createElement")
private external fun js_Document_createElement(
    arena: Arena,
    index: Object,
    tagPtr: Pointer,
    tagLen: Int,
    resultArena: Arena
): Object

@SymbolName("kvdom_Document_createTextNode")
private external fun js_Document_createTextNode(
    arena: Arena,
    index: Object,
    textPtr: Pointer,
    textLen: Int,
    resultArena: Arena
): Object

@SymbolName("kvdom_Document_defaultView")
private external fun js_Document_defaultView(
    arena: Arena,
    index: Object
): Object

File: kvdom/kvdom-dom/src/wasm32Main/resources/bindings/Document.js
konan.libraries.push({
  kvdom_Global_getDocument: function(resultArena) {
    return toArena(resultArena, document);
  },
  kvdom_Document_getElementById: function(
      arena, obj, idPtr, idLen, resultArena) {
    let id = toUTF16String(idPtr, idLen);
    let result = kotlinObject(arena, obj).getElementById(id);
    return result !== null && result !== undefined ? toArena(resultArena,
        result) : -1;
  },
  kvdom_Document_createElement: function(
      arena, obj, tagPtr, tagLen, resultArena) {
    let tag = toUTF16String(tagPtr, tagLen);
    let result = kotlinObject(arena, obj).createElement(tag);
    return toArena(resultArena, result);
  },
  kvdom_Document_createTextNode: function(
      arena, obj, textPtr, textLen, resultArena) {
    let text = toUTF16String(textPtr, textLen);
    let result = kotlinObject(arena, obj).createTextNode(text);
    return toArena(resultArena, result);
  },
  kvdom_Document_defaultView: function(arena, obj) {
    let defaultView = kotlinObject(arena, obj).defaultView;
    return defaultView == null ? -1 : toArena(arena, defaultView);
  },
});